### PR TITLE
[DependencyInjection] Allow for synthetic services before compilation

### PR DESCRIPTION
**Pull request by ro0NL**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | comma-separated list of tickets fixed by the PR, if any |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

Separated from #19606 

This allows accessing synthetic services from `ContainerBuilder::get` before compilation, besides they are exposed while extension loading.

It covers 2 usecases;
- synthetic service + synthetic definition (like before, but getting the service is allowed)
- synthetic service + non-synthetic definition (new feature => `ServiceAwareDefinition` which covers both usecases as well)

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php [hunk 1]
```diff
@@ -52,6 +52,10 @@ public function process(ContainerBuilder $container)
                 $tmpContainer->addExpressionLanguageProvider($provider);
             }
 
+            foreach ($container->getSynthetics() as $id => $service) {
+                $tmpContainer->set($id, $service);
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php]: Concerning BC: all services set via `PrependExtensionInterface::prepend` are available while extension loading.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 2]
```diff
@@ -435,12 +431,19 @@ public function get($id, $invalidBehavior = ContainerInterface::EXCEPTION_ON_INV
...
+            }
         } catch (ServiceNotFoundException $e) {
-            if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE !== $invalidBehavior) {
-                return;
+            if (ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE === $invalidBehavior) {
+                throw $e;
             }
 
-            throw $e;
+            return;
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Concerning BC: both `throw` and `return` now happens before the deprecation error...

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 3]
```diff
@@ -435,12 +431,19 @@ public function get($id, $invalidBehavior = ContainerInterface::EXCEPTION_ON_INV
...
+        }
+
+        if (!$this->compiled) {
+            @trigger_error(sprintf('Calling %s() before compiling the container is deprecated for non-synthetic services since version 3.2 and will throw an exception in 4.0.', __METHOD__), E_USER_DEPRECATED);
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Updated message. Concerning BC: before this was triggered for each possible alias, now only once.

**Comment by Ener-Getick** [comment]
@ro0NL as https://github.com/symfony/symfony/pull/19608 is merged, I guess you can close this one ? :)

**Comment by ogizanagi** [comment]
@Ener-Getick : My PR was only removing "dead code", it doesn't un-deprecate getting a service instance from the `ContainerBuilder`.

But I'm not convinced we need something like the new `ServiceAwareDefinition`. We could simply consider any service set using `ContainerBuilder::set` should be a "building-time only" service and allow to get it, but won't be available after compilation. It feels weird and I'm not convinced by this "feature" neither, but currently services set directly in the `ContainerBuilder` are vanished after compilation anyway.

**Comment by ro0NL** [comment]
Most important is the logic of `ContainerBuider::get` + `MergeExtensionConfigurationPass::process` for now.

`ServiceAwareDefinition` is somewhat a shortcut for `set()` + `setDefinition`, so this can be left out yes. Consider it a poc :)

**Comment by ro0NL** [comment]
@ogizanagi the `ServiceAwareDefinition` allows for having the service+definition at the same time, as `set()` [overwrites](https://github.com/ro0NL/symfony/blob/c5fb4a30d38c3df8f7c76d603031c83a4901ef48/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L376) definitions for the same id.

So if you have a non-synthetic definition + the ready to use service, you are stuck now. Not sure how common that is, on the other hand #19606 is poc of exactly that.

**Comment by ogizanagi** [comment]
@ro0NL : My point is that you probably don't need having a service definition for services designed to be used during the container building time only (but is it even a good idea to depend on any kind of service in this phase ?). :/ 
I think that's the only thing needed for the use-case of exposing kernel & bundles "metadata".
Otherwise, either you'll get a service definition into the compiled container that will never be used at runtime, either you'll have to set the service twice (during and after the container is compiled).

Thus [my suggestion about exposing a `booting_kernel`](https://github.com/symfony/symfony/pull/19594#issuecomment-239312668) during the container build, which would simply be a `KernelInterface` decorator throwing exception for unsafe methods, and will never be available at runtime (use the `kernel` service then). ^^

**Comment by ro0NL** [comment]
Could be.. this is just a generic approach that allows for it. If the component really needs it, i dont know. Perhaps nice to have. But yeah, my usecase is definitely kernel+bundles ;-)

A `booting_kernel` service will also do :+1:

**Comment by ogizanagi** [comment]
@ro0NL : regarding the part I've added after editing my comment above, here is a straightforward implementation of a contextualized container builder, allowing to share some context during the build time, before compilation of the container: https://github.com/symfony/symfony/compare/master...ogizanagi:contextualized_container

Here, the `Context` is locked right before calling `Compiler::compile`, ensuring nothing can mutate it appart from the `Kernel::getContainerBuilder()` method. But we can imagine (if legit) locking the context only after some point (in a dedicated compiler pass), making the context mutable at some point (extensions) but "safe to use" only in compiler passes.

Of course, I don't expect this to be a perfect nor legit implementation, but simply exploring and giving more hints on how to proceed considering your uses cases. :)

**Comment by ogizanagi** [comment]
> How does this reflect ContainerBuilder::set? Should it even be allowed then?

Indeed, it should not. I guess it was the idea behind the deprecation, but it still allows to set synthetic services, which are vanished after compilation, and moreover not merged during build phase. So it may not allow calling `set` at all.

> ContainerBuilder::getContext():Container

Yeah...thought about that too, but I'm not fan of giving visibility to the fact we use a container to build a container... 😅 At least the foolish lockable `Context` implementation hides that fact a little.

**Comment by ro0NL** [comment]
I dont mind :) I think we're too much used to _one_ container being the almighty thing. It's a pattern, eventually it's legit :wink: 

Another, bridge too far, idea; each bundle defines it's own container(s) in the ecosystem. Container as a service :scream:

Anyway, maybe focus this PR on deprecating `set()` first then? So at least the current behavior is consistent and clear.

**Comment by ogizanagi** [comment]
> Anyway, maybe focus this PR on deprecating set() first then? So at least the current behavior is consistent and clear.

IMHO it should be the subject of another PR deprecating the whole `set` usage.

**Pull request closed without merging**